### DISCUSSION TITLE
Performance lag due to Supabase getUser calls in Next.js middleware

### DISCUSSION THREAD

**Discussion by filigreti** [supabase/supabase#20905] [Questions] [answered]
# Bug report

- [y] I confirm this is a bug with Supabase, not with my own application. 
- [y] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

Calling supabase.auth.getUser() using the createServerClient with the @supabase/ssr package on every page route in Next.js middleware causes noticeable lag, especially on slow networks.
This negatively affects user experience by slowing down route transitions.

## To Reproduce

Create a Next.js app with middleware as shown in the supabase documentation.
Integrate Supabase with middleware as configured.
Navigate between routes, observing delays.
Turn on the browser slow mode to simulate slow network network speed

## Expected behavior

I expected the route transitions to be smooth and not experience noticeable lag, even on slower network connections.

## System information

- OS: [macOS]
- Browser (if applies) [hrome]
- Version of supabase-js: ["^2.39.3"]
- Version of Node.js: [stable version]

**Comment by charislam** [comment, marked as answer]
Hi @filigreti! Moved this to a discussion because the instructions and SDK work as intended, but there are certain mitigations you can try:

1. If you have routes that don't require a Supabase call, you can update the [matcher](https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher) to exclude them.
2. You can also validate the JWT yourself in middleware, rather than calling the Supabase Auth server to do it. We generally don't recommend this because the performance benefit may not be worth the security risks. It is **very important** that you:

- _Validate_ the JWT by checking the signature. _Do not_ just rely on what the cookie tells you, because anyone can send any cookie to you that they want.
- _Never_ leak the JWT signing key on the client. This can be easy to overlook on Next.js because of the mix of client and server code. Make sure wherever/however your signing key is imported, it is _always_ isolated to server-only code.

Any route that requires a Supabase call (i.e., requires going through the middleware) is already making that call (and probably other calls) to begin with, so there are probably other reasons causing the slowdown as well. You can look into ways to increase the perceived speed of the response, for example via transitions

**Reply by filigreti** [reply]
i have a /dashboard route that has multiple routes, the middleware service checks each route within the /dashboard, and in turn makes an api call to check when navigating each route..., if i remove the supabase middleware.. the app moves as expected.. with no issues

**Reply by justinianoleo** [reply]
Hi. Did you solve the problem? I'm having the same problem with my dashboard.

**Reply by ishaanbedi** [reply]
Hey @justinianoleo did you find a solution?

**Reply by justinianoleo** [reply]
@ishaanbedi Hi. No, i'm finding out a solution. Notice me if you know something.

**Reply by justinianoleo** [reply]
My solution was adding its own loading component to each page. I don't know if that's the best solution but i don't have that lag anymore.

**Reply by jacktsin1** [reply]
You could use `.auth.getSesion()`, extract the user object from the session, and make the necessary checks which will cover 99.9% of scenarios (requests made by legit users, blocking users that are currently logged out i.e.), and will have no impact on performance. 

For the tables that you want to protect you can use RLS (which will work for every req — an attacker may hit your API route directly which may not be covered by your middleware), so even when a user spoofs req.cookies to bypass the middleware's security layer it will be blocked by RLS which either way will authenticate the JWT from cookies of the user making the request if you specify a policy for the `authenticated` role or when calling `auth.uid()` / `auth.jwt()` to access the user's cokies.

For routes that perform other actions (i.e. communication with Stripe, etc.) that require the user to be authenticated, you could call `.auth.getUser()` individually on these specific routes together with additional security logic.

The only problem with that approach is that currently, supabase refreshes the user's session only when calling `.auth.getUser()` (not when calling `.auth.getSession()`), and thus if a user returns to your app 2 days later, they will be logged out.

So I would like to ask @charislam and the Supabase team first of all if I'm missing something in the approach above (security, performance), and moreover, what would be the recommended way to refresh a session without having to use `.getUser()` in the middleware which will affect performance.

**Reply by jacktsin1** [reply]
@kangmingtay @charislam Any updates about this?

**Reply by j4w8n** [reply]
Be careful with `getSession()` on the server side: https://github.com/orgs/supabase/discussions/23224

I heard the team was considering asymmetric JWTs and possibly using a caching mechanism for `getUser()` to reduce latency. I'm not sure if this is being worked on or if it's off the table.

**Reply by Kylar13** [reply]
> Be careful with `getSession()` on the server side: https://github.com/orgs/supabase/discussions/23224
> 
> I heard the team was considering asymmetric JWTs and possibly using a caching mechanism for `getUser()` to reduce latency. I'm not sure if this is being worked on or if it's off the table.

Any update on this? Is there any way we can make use of the new `waitUntil` to offload the cookie refresh?

**Reply by j4w8n** [reply]
@Kylar13 I'm not familiar with `waitUntil`.

**Reply by zineanteoh** [reply]
> Be careful with `getSession()` on the server side: https://github.com/orgs/supabase/discussions/23224
> 
> I heard the team was considering asymmetric JWTs and possibly using a caching mechanism for `getUser()` to reduce latency. I'm not sure if this is being worked on or if it's off the table.

both approaches seem to have downsides: 
approach 1: use `getSession()` and [risk security breaches](https://github.com/orgs/supabase/discussions/23224).
approach 2: use `getUser()` and [risk degraded performance](https://github.com/orgs/supabase/discussions/20905#discussion-6162769)

what is the recommended way then? 

(unfortunately all my existing routes have to go through middleware)

**Reply by kangmingtay** [reply]
> I heard the team was considering asymmetric JWTs and possibly using a caching mechanism for getUser() to reduce latency. I'm not sure if this is being worked on or if it's off the table.

We're working on this right now - if you need to verify the JWT to use the claims outside of Supabase, you need to perform the verification on your own.

Once we have support for asymmetric JWTs, it will be much easier (and safer) to support this natively in the supabase-js package.

**Reply by j4w8n** [reply]
> both approaches seem to have downsides: approach 1: use `getSession()` and [risk security breaches](https://github.com/orgs/supabase/discussions/23224). approach 2: use `getUser()` and [risk degraded performance](https://github.com/orgs/supabase/discussions/20905#discussion-6162769)
> 
> what is the recommended way then?
> 
> (unfortunately all my existing routes have to go through middleware)

Supabase official recommendation is to call `getUser`. It might be best to stick with this for a bit longer, until the work kangmingtay spoke of is finished.

If you're willing to take things into your own hands, then the non-recommended but lowest-latency and safe way (in my opinion) is to verify and decode the JWT and create your own validated session. Be sure to only use your project's JWT secret on the server-side (which I heard can be a bit trickier in Next?). I have a SvelteKit example of this, but it's just js/ts if you're using a different framework. After the the team is done with their work, I plan to reevalute my approach. Checkout the `event.locals.getSession` code in https://github.com/j4w8n/sveltekit-supabase-ssr/blob/main/src/hooks.server.ts

**Reply by pppdns** [reply]
@kangmingtay any updates on the asymetric JWT work? Is there some branch/issue that we can subscribe to to get updates? (and possibly contribute?)

**Reply by CesGalaxy** [reply]
> The only problem with that approach is that currently, supabase refreshes the user's session only when calling `.auth.getUser()` (not when calling `.auth.getSession()`), and thus if a user returns to your app 2 days later, they will be logged out.

I hope this helps someone:

My app was very slow with the middleware and I managed to fix it by creating an API route that calls the `getUser()` method, then at the root layout I wrapped all the pages with a client component that called that API route (I think you could make it run just once a day, so it doesn't call the API without reason) and the cookies get updated after refreshing the user token.

Sorry for my bad english :P

**Reply by PatrickRogg** [reply]
@kangmingtay any idea when you or someone from the team get a chance to work on this or when it will be completed?

I think this is one of the most impactful changes that you could make for Next.js. 

I tested both `getUser` and `getSession` and the difference is massive. You literally can feel the difference as a user when navigating through pages. With `getUser` switching pages feels clunky and slow. With `getSession` switching pages is almost instant.

**Reply by lukekim** [reply]
@charislam @kangmingtay can we re-open this discussion/issue? This significantly affects us to the degree we're considering migrating away from Supabase. Do you have any updates on the asymmetric JWT + caching work?

For transparency I posted about this on X: https://x.com/0xLukeKim/status/1883948488788197423

**Reply by louisthomaspro** [reply]
Hi @lukekim,
Meanwhile, I suggest you the solution of @anarkrypto => https://github.com/orgs/supabase/discussions/23224#discussioncomment-10082128

```js
const safeSession = new SupabaseSafeSession(supabase, process.env.SUPABASE_JWT_SECRET!)
const { data: user, error } = await safeSession.getUser()
```

I use it for the middleware and server components.

**Reply by o1Suleyman** [reply]
> Supabase official recommendation is to call `getUser`. It might be best to stick with this for a bit longer, until the work kangmingtay spoke of is finished.

well that aged terribly. It's February 2025 and we still have nothing

**Reply by o1Suleyman** [reply]
> Hi @lukekim, Meanwhile, I suggest you the solution of @anarkrypto => https://github.com/orgs/supabase/discussions/23224#discussioncomment-10082128
> 
> ```js
> const safeSession = new SupabaseSafeSession(supabase, process.env.SUPABASE_JWT_SECRET!)
> const { data: user, error } = await safeSession.getUser()
> ```
> 
> I use it for the middleware and server components.

what does it do?

**Reply by anarkrypto** [reply]
@o1Suleyman it allows you to verify the user session safely on your backend without calling getUser (which makes an api request bringing performance issue).

**Comment by tadeumaia** [comment]
https://supabase.com/blog/jwt-signing-keys as of Jul 2025. as I always got myself into this thread, this might help someone else.
